[WebDriver][BiDi] Add support to relay commands with new processBidiMessage and bidiMessageSent in Source/WebDriver

lauromoura opened the pull request:
#### 4b386b5e7393fbda26d42c08f9cc38f2a6fff81b
<pre>
[WebDriver][BiDi] Add support to relay commands with new processBidiMessage and bidiMessageSent in Source/WebDriver
<a href="https://bugs.webkit.org/show_bug.cgi?id=288194">https://bugs.webkit.org/show_bug.cgi?id=288194</a>

Reviewed by BJ Burg.

290582@main added two new features in Automation.json:

- processBidiMessage command
- bidiMessageSent event

This pair can be used to relay BiDi commands to the browser, which in
turn can send almost ready BiDi responses and events. In this scenario,
the service acts basically as a proxy, instead of worrying with
individual commands details.

Another benefit of this is that we can move more code inside WebKit itself,
allowing the ports not using `Source/WebDriver` directly to share more code,
and keeping the service code simpler.

This commit removes the previous `Automation.json` `logEntryAdded`
event, relying on the new `bidiMessageSent` event to wrap it. In a
follow-up commit, we&apos;ll move the post-processing code in
`Session::doLogEntryAdded` back inside `WebAutomationSession` so the
service can receive the ready event body.

This commit also makes WebDriverBidiProcessor translate &quot;inspector
protocol-like&quot; errors into the WebDriver-BiDi messages, so the driver
can just forward it ahead.

* Source/WebDriver/Session.cpp:
(WebDriver::Session::Session): Rename EventHandler to BiDiHandler
(WebDriver::Session::dispatchBiDiMessage): Renamed from dispatch event
(WebDriver::Session::doLogEntryAdded): Better logging and no need to
check if event is enabled, as we do it in doBidiMessageSent.
(WebDriver::Session::enableGlobalEvent): No need to strip the &apos;.&apos;
between the module and event name, as they are not part of the
Automation.json protocol anymore.
(WebDriver::Session::disableGlobalEvent): Ditto.
(WebDriver::Session::relayBidiCommand): Forward the incoming bidi
message to the browser, replying the incoming message with the returned
payload.
(WebDriver::Session::doBidiMessageSent): Added. Unpack the received Bidi
message and process either the command reply or the incoming event.
(WebDriver::Session::dispatchEvent): Deleted.
(WebDriver::Session::toInternalEventName): Deleted.
* Source/WebDriver/Session.h: Rename EventHandler to BiDiMessageHandler
* Source/WebDriver/SessionHost.cpp:
(WebDriver::SessionHost::dispatchMessage): Ditto.
(WebDriver::SessionHost::dispatchBiDiMessage): Added.
(WebDriver::SessionHost::dispatchEvent): Deleted.
* Source/WebDriver/SessionHost.h: Rename EventHandler to
  BiDiMessageHandler
* Source/WebDriver/WebDriverService.cpp:
(WebDriver::WebDriverService::handleMessage): Fallback to
relayBidiCommand when we don&apos;t have a local method impl.
(WebDriver::WebDriverService::findBidiCommand): Simplify parameter
passing.
(WebDriver::WebDriverService::bidiSessionSubscribe): Logging.
* Source/WebDriver/WebDriverService.h:
* Source/WebKit/UIProcess/Automation/Automation.json: Remove uneeded
  logEntryAdded event.
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp: Ditto.
(WebKit::WebAutomationSession::logEntryAdded):
* WebDriverTests/TestExpectations.json: Drive-by bug URL fix.

* Source/WebDriver/Session.cpp:
(WebDriver::Session::Session): Rename EventHandler to BidiMessageHandler
(WebDriver::Session::dispatchBidiMessage): Renamed from dispatchEvent.
This function validates the message to be forwarded to the browser.
(WebDriver::Session::doLogEntryAdded): Move event enabled check
outside.
(WebDriver::Session::enableGlobalEvent): No need to strip the &apos;.&apos;
between the module and event name, as they&apos;re wrapped in
&apos;bidiMessageSent&apos; now.
(WebDriver::Session::disableGlobalEvent): Ditto.
(WebDriver::Session::relayBidiCommand): Forward the incoming bidi
message to the browser, with the callback just observing for errors
delivering the request.
(WebDriver::Session::dispatchEvent): Deleted.
(WebDriver::Session::toInternalEventName): Deleted.
* Source/WebDriver/Session.h: Rename EventHandler to BidiMessageHandler
* Source/WebDriver/SessionHost.cpp:
(WebDriver::SessionHost::dispatchMessage): Ditto.
(WebDriver::SessionHost::dispatchBidiMessage): Replaces dispatchEvent.
(WebDriver::SessionHost::dispatchEvent): Deleted.
* Source/WebDriver/SessionHost.h: Rename EventHandler to
  BidiMessageHandler.
* Source/WebDriver/WebDriverService.cpp:
(WebDriver::WebDriverService::handleMessage): Fallback to
relayBidiCommand when we don&apos;t have a local method impl.
(WebDriver::WebDriverService::findBidiCommand): Simplify parameter
passing.
* Source/WebDriver/WebDriverService.h:
* Source/WebDriver/WebSocketServer.cpp:
(WebDriver::WebSocketServer::sendMessage): Added helpers around getting
the connection for a given SessionId.
(WebDriver::WebSocketServer::sendErrorResponse): Ditto.
* Source/WebDriver/WebSocketServer.h:
* Source/WebDriver/soup/WebSocketServerSoup.cpp:
(WebDriver::WebSocketServer::sendMessage): No need to iterate the
available sessions.
* Source/WebKit/UIProcess/Automation/Automation.json: Removed deprecated
  logEntryAdded event, relying on bidiMessageSent instead.
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::logEntryAdded): Ditto.
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
(WebKit::toBiDiErrorCode): Added, to help translate our inspector-based
errors into BiDi errors.
(WebKit::WebDriverBidiProcessor::sendBidiMessage): Add &quot;type&quot; field in
bidi message and translate the error message to BiDi ErrorResponse
production.
* WebDriverTests/TestExpectations.json: Drive-by gardening of W3C BiDi
  tests.

Canonical link: <a href="https://commits.webkit.org/294391@main">https://commits.webkit.org/294391@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3344770791e1258bd83576e28887963441af158e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100944 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20606 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10909 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106090 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51543 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20914 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29100 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76871 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33901 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103951 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16084 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91174 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57219 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15899 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9196 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50918 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85807 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9257 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108446 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28072 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20643 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85841 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28434 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87374 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85383 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21875 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30090 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7818 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/22105 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28002 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33270 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27814 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31134 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29372 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

burg reviewed (commented):
Commit typo: 'logging.and'

burg commented on Source/WebDriver/Session.cpp:
```diff
@@ -3152,21 +3153,18 @@ void Session::takeScreenshot(std::optional<String> elementID, std::optional<bool
 
 #if ENABLE(WEBDRIVER_BIDI)
-void Session::dispatchEvent(RefPtr<JSON::Object>&& message)
+void Session::dispatchBiDiMessage(RefPtr<JSON::Object>&& message)
```
For the Source/WebKit code we have been spelling it as `Bidi`, so I'd prefer to maintain that pattern here.

lauromoura replied:
I'd be fine with changing to Bidi, but <bikeshedding>given the protocol name is with a capital `D` (`BiDi`) and WebKit already uses `Bidi` for "bidirectional" text, maybe should we change `Source/WebKit` to `BiDi` instead? The case difference could help </bikeshedding?>

In any case, I'll leave changing other occurrences to a separate commit.

burg replied:
I think Bidi is appropriate in both places because it means bidirectional in both cases. Just because bidirectional text exists, doesn't mean we can't use the term elsewhere. They are rarely confused in context.

The WebKit style guide is inconclusive on `BiDi` vs `Bidi`: https://webkit.org/code-style-guidelines/#names-basic

burg commented on Source/WebDriver/Session.cpp:
```diff
@@ -3152,21 +3153,18 @@ void Session::takeScreenshot(std::optional<String> elementID, std::optional<bool
-        doLogEntryAdded(WTFMove(message));
+    auto messageName = method.substring(automationPrefix.length());
+    if (messageName == "bidiMessageSent"_s) {
+        doBidiMessageSent(WTFMove(message));
```
naming nit: prefer something like `dispatchBidiMessage` or `handleBidiMessage` to reflect that this is dispatching a message to handlers.

lauromoura replied:
Actually, I'm going to inline this patch's new `doBiDiMessageSent` inside this current function, which is already very short and called `dispatchBiDiMessage`

burg commented on Source/WebDriver/Session.h:
```diff
@@ -159,7 +159,8 @@ public RefCounted<Session>
     void disableGlobalEvent(const String&);
-    void dispatchEvent(RefPtr<JSON::Object>&&);
+    void dispatchBiDiMessage(RefPtr<JSON::Object>&&);
+    void relayBidiCommand(const String&, unsigned commandId, Function<void(WebSocketMessageHandler::Message&&)>&&);
```
Might be possible to use `CompletionHandler` instead of `Function` here. You'd need to clear out `m_relayedCommands` in the destructor to avoid asserting.

lauromoura replied:
I went with `Function<...>` because, IIUC, the spec does not require replying to pending commands when the session ends, either through `session.end()`[1] or `browser.close()`[2]. In both cases, the client is notified by the closure of the WebSocket connection.

[1] https://w3c.github.io/webdriver-bidi/#commands-sessionend
[2] https://w3c.github.io/webdriver-bidi/#command-browser-close

burg commented on Source/WebDriver/Session.h:
```diff
@@ -275,12 +276,16 @@ public RefCounted<Session>
     HashSet<String> m_globalEventSet;
     WeakPtr<WebSocketServer> m_bidiServer;
 
+    // Relayed messages
```
Naming nit: `m_pendingCommands` or `m_pendingRelayedCommands`

lauromoura replied:
I actually removed this, as we don't need to keep track of the completion handler given the `WebSocketServer` maps the connection to a given session.

burg commented on Source/WebDriver/WebDriverService.h:
```diff
@@ -134,7 +134,7 @@ class WebDriverService final : public HTTPRequestHandler
 
 #if ENABLE(WEBDRIVER_BIDI)
-    // BiDi message handlers
+    // BiDi message handlers for methods that require some work on the driver side, without relaying direclty to the browser
```
Typo: directly

burg reviewed (commented):
This looks good for the case when commands succeed.

One thing that I found out the hard way, you'll need to do some special handling of Bidi messages from WebKit that are "protocol errors" inside `WebDriverBidiProcessor::bidiMessageSent`. If the bidi message doesn't parse as a command response or event, then it's a "protocol error". You'll need to remove the associated callback from `m_relayedCommands` to report the protocol error as a command failure.

yury-s commented on Source/WebDriver/WebDriverService.h:
```diff
@@ -134,7 +134,7 @@ class WebDriverService final : public HTTPRequestHandler
 
 #if ENABLE(WEBDRIVER_BIDI)
-    // BiDi message handlers
+    // BiDi message handlers for methods that require some work on the driver side, without relaying directly to the browser
```
Just drop this comment?

yury-s commented on Source/WebDriver/WebSocketServer.h:
```diff
@@ -129,7 +129,10 @@ class WebSocketServer : public RefCountedAndCanMakeWeakPtr<WebSocketServer> {
     String getSessionID(const String& resource);
-    void sendMessage(const String& session, const String& message);
+    void sendMessage(const String& sessionId, const String& message);
+    void sendMessage(WebSocketMessageHandler::Connection, const String& message);
```
Can the methods that accept `WebSocketMessageHandler::Connection` as a parameter be made private to `WebSocketServer`?

From what I can tell, they're primarily used by the platform-specific (Soup) implementation. `WebDriverService` and `WebSocketServer` currently communicate using  both `sessionId` and `connection` but it seems that `sessionId` alone would suffice and allow for better encapsulation of the server logic.

burg replied:
Agree with Yury's suggestion, except... As commented elsewhere, I prefer using `clientId` rather than `sessionId` for this purpose. For BiDi-only sessions that use the static session creation endpoint, they won't have a `sessionId` unless `session.new` succeeds. And even then, the client could send the `session.status` command which has no associated session.

lauromoura replied:
Moving the `Connection` to private is fine, I'll do it.

On the other hand, about `sessionId` vs `clientId`, that's a good point.

`session.status` is fine, because it replies through the completion handler directly, not passing through this method.

`session.new` on the other hand, will receive an async reply from the browser, but as you mentioned, its original `Connection` will not be mapped yet to any known `Session`.

Given most of the currently supported commands are handled through valid sessions, would it be ok to leave the `sessionId` in _this version_, and we add proper `clientId` support with static methods in a separate https://bugs.webkit.org/show_bug.cgi?id=281721?

yury-s commented on Source/WebDriver/Session.h:
```diff
@@ -159,7 +159,8 @@ public RefCounted<Session>
     void disableGlobalEvent(const String&);
-    void dispatchEvent(RefPtr<JSON::Object>&&);
+    void dispatchBidiMessage(RefPtr<JSON::Object>&&);
+    void relayBidiCommand(const String&, unsigned commandId, Function<void(WebSocketMessageHandler::Message&&)>&&);
```
The method doesn't need to know `commandId`, there is already 1-1 correlation between the command and its callback. The id is only needed for the websocket reply and could be added later by the caller.

lauromoura replied:
The issue here was that `relayBidiCommand` callback receives a `SessionHost::CommandResponse`, which is not used in `WebDriverService` (the caller). And as currently `WebSocketMessageHandler::Message` stores the messages as its String payload, it would be awkward for the caller to inject the command id.

I'll make `Message` store a regular JSON object, and we serialize it just when accessing the payload to send. That will make it easier for the caller to update individual message fields.

lauromoura replied:
Actually, I think it's better to make this `Message` change on a separate patch, as it might grow a bit.

burg commented on Source/WebDriver/Session.cpp:
```diff
@@ -3152,31 +3154,67 @@ void Session::takeScreenshot(std::optional<String> elementID, std::optional<bool
     auto method = message->getString("method"_s);
-    if (!method.startsWith(automationPrefix)) {
-        WTFLogAlways("Unknown event domain: %s", method.utf8().data());
+    if (method == "Automation.browsingContextCleared"_s)
```
To make this method be *only* about dispatching the Bidi message, I feel like these first two checks should be in the caller. The call stack would be something like:

Session::dispatchMessage() -> dispatchEvent() -> dispatchBidiMessage() -> bidiServer->sendMessage()

lauromoura replied:
Ok. In this case, as I renamed `dispatchEvent` to `dispatchBidiMessage`, I'll move the checks to `dispatchMessage`.

burg commented on Source/WebDriver/SessionHost.h:
```diff
@@ -97,7 +97,7 @@ class SessionHost final
 
 #if ENABLE(WEBDRIVER_BIDI)
-    void addEventHandler(WeakPtr<BiDiEventHandler>&& handler) { m_eventHandler = WTFMove(handler); }
+    void addBidiHandler(WeakPtr<BidiMessageHandler>&& handler) { m_bidiHandler = WTFMove(handler); }
```
If there is only one, then it should be `setBidiHandler`. I'd prefer `setBidiMessageHandler` to match the type.

burg commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -465,7 +465,7 @@ void WebDriverService::handleMessage(WebSocketMessageHandler::Message&& message,
     auto parsedMessageValue = JSON::Value::parseJSON(String::fromUTF8(message.payload.data()));
     if (!parsedMessageValue) {
-        RELEASE_LOG(WebDriverBiDi, "WebDriver handle Message: Failed to parse incoming message");
+        RELEASE_LOG(WebDriverBiDi, "WebDriver handle BiDi message: Failed to parse incoming message");
```
Nit: prefer to use the method name here:

`RELEASE_LOG(WebDriverBiDi, "WebDriverService::handleMessage() failed to parse BiDi message as JSON");`

burg commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -475,46 +475,42 @@ void WebDriverService::handleMessage(WebSocketMessageHandler::Message&& message,
         return;
     }
 
+    const auto& asObject = parsedMessageValue->asObject();
```
Nit: `messageObject`

burg commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -2494,9 +2495,12 @@ static String logEntryTypeForMessage(const JSC::MessageSource& messageSource)
 }
+#endif // ENABLE(WEBDRIVER_BIDI)
 
+// FIXME We can guard this function too once we create a feature flag to control it
```
Nit: comment isn't necessary

burg commented on Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
```diff
@@ -25,6 +25,8 @@
 
 #include "config.h"
 #include "WebDriverBidiProcessor.h"
+#include "AutomationBackendDispatchers.h"
```
Style: add an extra newline after the primary header include

burg commented on Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
```diff
@@ -94,6 +96,77 @@ void WebDriverBidiProcessor::processBidiMessage(const String& message)
+    case Inspector::Protocol::Automation::ErrorMessage::NoSuchExtension:
+        return "no such web extension"_s;
+    default:
+        return "uknown error"_s;
```
Nit: unknown

burg commented on Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
```diff
@@ -94,6 +96,77 @@ void WebDriverBidiProcessor::processBidiMessage(const String& message)
+    if (!errorMessage)
+        return "unknown error"_s;
+
+    // Translation originally based on the one in Source/WebDriver/CommandResult.cpp
```
Comment not needed, maybe move it to the changelog.

burg commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -475,46 +475,42 @@ void WebDriverService::handleMessage(WebSocketMessageHandler::Message&& message,
+    if (!findBidiCommand(asObject, &handler, parameters)) {
+        RELEASE_LOG(WebDriverBiDi, "Failed to find appropriate BiDi command on WebDriver service. Relaying to the browser.");
+        m_session->relayBidiCommand(makeString(message.payload), *commandId, [completionHandler = WTFMove(completionHandler), message](WebSocketMessageHandler::Message&& resultMessage) {
+            resultMessage.connection = message.connection;
```
It seems a little weird to me that we keep the client context by setting the connection on each `resultMessage`. If `connection` was wrapping some networking library objects, I would be concerned about leaking those objects if we accidentally retained or copied resultMessage. What we really want to know is which *websocket* client originated the command. I would prefer that the WebSocketServer vend an opaque `ClientIdentifier` to represent WebSocket clients instead of connection objects.

lauromoura replied:
Good point. I'll modify to save just the `sessionID` associated to the request in the closure instead, given we can get the associated connection inside the callback.

As for `clientID`, we can add a proper scheme when adding static methods, as I mentioned above. WDYT?

burg commented on Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
```diff
@@ -105,7 +178,40 @@ void WebDriverBidiProcessor::sendBidiMessage(const String& message)
     LOG(Automation, "%s", message.utf8().data());
 
-    session->sendBidiMessage(message);
+    auto msgValue = JSON::Object::parseJSON(message);
```
This code is performing the correct operations to unpack the Inspector protocol message.. but I believe it should be performed in WebDriverService or WebDriverServer where the `Automation.bidiMessageSent` event is handled. Rather than on the UIProcess side.

If we construct the object here, we won't have the original id supplied by the client when the command was first received, right?

lauromoura replied:
> If we construct the object here, we won't have the original id supplied by the client when the command was first received, right?

The complete message is being forwarded as the payload of `Automation.json`'s `processBidiMessage`, including its command Id. `WebDriverProcessor::sendBidiMessage` in turn is getting a reply/error response with the right command ID already.

But, at a closer look, this is kinda working "by accident", by making the bidi message id also work as the ID in the Bidi domain dispatchers, right? And it would break if a client ever uses something like non-integer ids.

yury-s commented on Source/WebDriver/Session.cpp:
```diff
@@ -3261,35 +3299,29 @@ bool Session::eventIsEnabled(const String& eventName, const Vector<String>&)
+    parameters->setString("message"_s, message);
+    m_host->sendCommandToBackend("processBidiMessage"_s, WTFMove(parameters), [protectedThis = Ref { *this }, completionHandler = WTFMove(completionHandler), commandId](SessionHost::CommandResponse&& response) {
+        if (response.isError) {
+            auto errorCode = CommandResult::ErrorCode::UnknownError;
```
With the current implementation, this should map all errors returned by WebDriverBidiProcessor to "unknown error". Maybe add a FIXME for now, as I think properly plumbing the bidi errors would require quite some work, including changes to the dispatchers (as discussed on Slack).

lauromoura replied:
In a way, yes, but this response block is related to problems when dispatching the command. I mean, problems before reaching `WebDriverBidiProcessor::processBidiMessage(message)`, not errors related to the command execution.

The latter would come as separate messages, handled by `sendBidiMessage`.

Here's a sequence diagram related to it. In this case, any error after `WebDriverBidiProcessor::processBiDi` will be reported as a separate message, the bottom half of the diagram.

```mermaid
sequenceDiagram

participant Client
participant WebDriverService

box Source/WebKit
participant WebAutomationSession
participant WebDriverBiDiProcessor
participant BackendDispatcher
end

box Generated Code
participant WebDriverBidiBackendDispatcher
end


note over WebDriverBidiBackendDispatcher: Generated from Bidi*.json protocols

Client ->> WebDriverService: webSocket(bidiJSON)

WebDriverService ->> WebAutomationSession: processBiDiMessage(bidiJSON)

note right of WebAutomationSession: Command dispatch (maybe async)
rect rgba(0, 0, 255, 0.1) 
WebAutomationSession ->> WebDriverBiDiProcessor: processBiDiMessage(bidiJSON)
WebDriverBiDiProcessor ->> BackendDispatcher: dispatch(bidiJSON)
BackendDispatcher ->> BackendDispatcher: parseJSON(bidiJSON)
BackendDispatcher ->> WebDriverBidiBackendDispatcher: dispatch(parsedJSON)
WebDriverBidiBackendDispatcher ->> WebDriverBidiBackendDispatcher: navigate(parsedJSON)
WebDriverBidiBackendDispatcher ->> WebDriverBidiBackendDispatcher: parseParams(parsedJSON)
WebDriverBidiBackendDispatcher ->> WebDriverBiDiProcessor: navigate(CommandParams, Callback)

WebDriverBiDiProcessor ->> WebDriverBiDiProcessor: waitNavigation()

end
WebAutomationSession -->> WebDriverService: VoidReply()
note right of WebDriverService: Handle problems<br>delivering the message

note right of WebAutomationSession: Command Reply
rect rgba(0, 0, 255, 0.1) 
WebDriverBiDiProcessor -->> WebDriverBidiBackendDispatcher: Callback(CommandResult)
WebDriverBidiBackendDispatcher ->> BackendDispatcher: sendResponse(inspectorJSON)
BackendDispatcher ->> WebDriverBiDiProcessor: sendMessageToFrontend(inspectorJSON)
WebDriverBiDiProcessor ->> WebDriverBiDiProcessor: sendBidiMessage(inspectorJSON)
note right of WebDriverBiDiProcessor: Translate inspector fields into bidi fields

WebDriverBiDiProcessor ->> WebAutomationSession: sendBidiMessage(bidiJSON)
end
WebAutomationSession ->> WebDriverService: bidiMessageSent(bidiJSON)
WebDriverService -->> Client: webSocket(bidiJSON)
```

yury-s replied:
Yeah, makes sense. I spotted the inadvertent mapping to UnknownError while looking the browser module tests, but it turned out to be due to the `case -32000: // Server error` [problem](https://github.com/WebKit/WebKit/pull/42968#discussion_r2056691991) in the processor that you've already addressed.

yury-s commented on Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
```diff
@@ -94,6 +96,77 @@ void WebDriverBidiProcessor::processBidiMessage(const String& message)
+    case -32700: // Parse error
+    case -32600: // Invalid request
+    case -32603: // Internal error
+    case -32000: // Server error
```
All handler errors are server errors, which means that the control flow will never reach parseEnumValueFromString below. You probably need to move the error mapping into this "Server error" branch.

This entire approach to error handling feels overcomplicated, we'll need to address it separately.

lauromoura replied:
Nice catch, this is what `CommandResult` does for the classic errors and I missed it while testing.

And I agree the error handling will need some follow-ups changes, as this version is kinda "monkey patching" on top of the classic error handling.

yury-s commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -475,46 +475,42 @@ void WebDriverService::handleMessage(WebSocketMessageHandler::Message&& message,
-        auto errorCode = CommandResult::ErrorCode::UnknownCommand;
-        auto errorReply = WebSocketMessageHandler::Message::fail(errorCode, connection, { "Command not supported"_s }, commandId);
-        completionHandler(WTFMove(errorReply));
+    if (!findBidiCommand(asObject, &handler, parameters)) {
```
Message parsing and the handler dispatcher can be generated from `BidiSession.json` similar to other Bidi modules, you shouldn't need to keep a static list of methods. Maybe add a FIXME for now?

lauromoura replied:
Yes, in the long run we could add similar dispatchers here too. It's a leftover from the original bidi prototype, based on the classic implementation. I'll link the static methods issue in the comment (https://bugs.webkit.org/show_bug.cgi?id=281721)

yury-s reviewed (approved):
Looks good to me (not a reviewer). Let’s merge it to unblock Bidi testing on Linux, we can clean things up in follow-ups.

lauromoura committed:
[WebDriver][BiDi] Add support to relay commands with new processBidiMessage and bidiMessageSent in Source/WebDriver
https://bugs.webkit.org/show_bug.cgi?id=288194

Reviewed by BJ Burg.

290582@main added two new features in Automation.json:

- processBidiMessage command
- bidiMessageSent event

This pair can be used to relay BiDi commands to the browser, which in
turn can send almost ready BiDi responses and events. In this scenario,
the service acts basically as a proxy, instead of worrying with
individual commands details.

Another benefit of this is that we can move more code inside WebKit itself,
allowing the ports not using `Source/WebDriver` directly to share more code,
and keeping the service code simpler.

This commit removes the previous `Automation.json` `logEntryAdded`
event, relying on the new `bidiMessageSent` event to wrap it. In a
follow-up commit, we'll move the post-processing code in
`Session::doLogEntryAdded` back inside `WebAutomationSession` so the
service can receive the ready event body.

This commit also makes WebDriverBidiProcessor translate "inspector
protocol-like" errors into the WebDriver-BiDi messages, so the driver
can just forward it ahead.

* Source/WebDriver/Session.cpp:
(WebDriver::Session::Session): Rename EventHandler to BiDiHandler
(WebDriver::Session::dispatchBiDiMessage): Renamed from dispatch event
(WebDriver::Session::doLogEntryAdded): Better logging and no need to
check if event is enabled, as we do it in doBidiMessageSent.
(WebDriver::Session::enableGlobalEvent): No need to strip the '.'
between the module and event name, as they are not part of the
Automation.json protocol anymore.
(WebDriver::Session::disableGlobalEvent): Ditto.
(WebDriver::Session::relayBidiCommand): Forward the incoming bidi
message to the browser, replying the incoming message with the returned
payload.
(WebDriver::Session::doBidiMessageSent): Added. Unpack the received Bidi
message and process either the command reply or the incoming event.
(WebDriver::Session::dispatchEvent): Deleted.
(WebDriver::Session::toInternalEventName): Deleted.
* Source/WebDriver/Session.h: Rename EventHandler to BiDiMessageHandler
* Source/WebDriver/SessionHost.cpp:
(WebDriver::SessionHost::dispatchMessage): Ditto.
(WebDriver::SessionHost::dispatchBiDiMessage): Added.
(WebDriver::SessionHost::dispatchEvent): Deleted.
* Source/WebDriver/SessionHost.h: Rename EventHandler to
  BiDiMessageHandler
* Source/WebDriver/WebDriverService.cpp:
(WebDriver::WebDriverService::handleMessage): Fallback to
relayBidiCommand when we don't have a local method impl.
(WebDriver::WebDriverService::findBidiCommand): Simplify parameter
passing.
(WebDriver::WebDriverService::bidiSessionSubscribe): Logging.
* Source/WebDriver/WebDriverService.h:
* Source/WebKit/UIProcess/Automation/Automation.json: Remove uneeded
  logEntryAdded event.
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp: Ditto.
(WebKit::WebAutomationSession::logEntryAdded):
* WebDriverTests/TestExpectations.json: Drive-by bug URL fix.

* Source/WebDriver/Session.cpp:
(WebDriver::Session::Session): Rename EventHandler to BidiMessageHandler
(WebDriver::Session::dispatchBidiMessage): Renamed from dispatchEvent.
This function validates the message to be forwarded to the browser.
(WebDriver::Session::doLogEntryAdded): Move event enabled check
outside.
(WebDriver::Session::enableGlobalEvent): No need to strip the '.'
between the module and event name, as they're wrapped in
'bidiMessageSent' now.
(WebDriver::Session::disableGlobalEvent): Ditto.
(WebDriver::Session::relayBidiCommand): Forward the incoming bidi
message to the browser, with the callback just observing for errors
delivering the request.
(WebDriver::Session::dispatchEvent): Deleted.
(WebDriver::Session::toInternalEventName): Deleted.
* Source/WebDriver/Session.h: Rename EventHandler to BidiMessageHandler
* Source/WebDriver/SessionHost.cpp:
(WebDriver::SessionHost::dispatchMessage): Ditto.
(WebDriver::SessionHost::dispatchBidiMessage): Replaces dispatchEvent.
(WebDriver::SessionHost::dispatchEvent): Deleted.
* Source/WebDriver/SessionHost.h: Rename EventHandler to
  BidiMessageHandler.
* Source/WebDriver/WebDriverService.cpp:
(WebDriver::WebDriverService::handleMessage): Fallback to
relayBidiCommand when we don't have a local method impl.
(WebDriver::WebDriverService::findBidiCommand): Simplify parameter
passing.
* Source/WebDriver/WebDriverService.h:
* Source/WebDriver/WebSocketServer.cpp:
(WebDriver::WebSocketServer::sendMessage): Added helpers around getting
the connection for a given SessionId.
(WebDriver::WebSocketServer::sendErrorResponse): Ditto.
* Source/WebDriver/WebSocketServer.h:
* Source/WebDriver/soup/WebSocketServerSoup.cpp:
(WebDriver::WebSocketServer::sendMessage): No need to iterate the
available sessions.
* Source/WebKit/UIProcess/Automation/Automation.json: Removed deprecated
  logEntryAdded event, relying on bidiMessageSent instead.
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::logEntryAdded): Ditto.
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
(WebKit::toBiDiErrorCode): Added, to help translate our inspector-based
errors into BiDi errors.
(WebKit::WebDriverBidiProcessor::sendBidiMessage): Add "type" field in
bidi message and translate the error message to BiDi ErrorResponse
production.
* WebDriverTests/TestExpectations.json: Drive-by gardening of W3C BiDi
  tests.

Canonical link: https://commits.webkit.org/294391@main
